Change azureblob backend to use new official client SDK

sandeepkru opened the pull request:
backend/azureblob: Port new Azure Blob Storage SDK #2362

This change includes removing older azureblob storage SDK, and getting parity to existing code with latest blob storage SDK.

This change is also pre-req for addressing #2091

ncw commented:
This is looking very good - well done :-)

Looking at the travis output it looks like that azblob isn't supported under these OSes.  You can test this locally with `make compile_all`.

(I added a patch to make the cross compile output a bit more readable 97beff5370ce791010f488e1729cde7650af534a - rebase to pick it up!)

```
2018/07/02 10:08:17 11 compile failures:
  freebsd/386
  freebsd/amd64
  freebsd/arm
  netbsd/386
  netbsd/amd64
  netbsd/arm
  openbsd/386
  openbsd/amd64
  plan9/386
  plan9/amd64
  solaris/amd64
```

So what I would do here is report an upstream bug and put in build constraints.  See the sftp backend for an example of the build constraints needed.

It also looks like that azblob doesn't support go1.7 both GetBody and NoBody are mentioned as being new in the [go1.8 release notes](https://golang.org/doc/go1.8).

```
# github.com/ncw/rclone/vendor/github.com/Azure/azure-pipeline-go/pipeline
vendor/github.com/Azure/azure-pipeline-go/pipeline/request.go:49: r.GetBody undefined (type Request has no field or method GetBody)
vendor/github.com/Azure/azure-pipeline-go/pipeline/request.go:58: undefined: http.NoBody
vendor/github.com/Azure/azure-pipeline-go/pipeline/request.go:59: r.GetBody undefined (type Request has no field or method GetBody)
vendor/github.com/Azure/azure-pipeline-go/pipeline/request.go:60: undefined: http.NoBody
vendor/github.com/Azure/azure-pipeline-go/pipeline/request.go:95: undefined: http.NoBody
vendor/github.com/Azure/azure-pipeline-go/pipeline/request.go:107: undefined: http.NoBody
```

I'd do the same, upstream bug and build constraint.

You'll end up with build constraints which look something like this I think

```
// +build !freebsd,!netbsd,!openbsd,!plan9,!solaris,go1.8
```

And for the _unsupported file (which doesn't exist yet, but you'll need it otherwise the builds will break).

```
// +build freebsd netbsd openbsd plan9 solaris !go1.8
```

I haven't attempted to review the code yet other than looking at the FIXMEs - I thought we should get it passing the tests first.  However I did note that in your vendor commit, can you prefix the line wiith `vendor:` rather than `backend/azureblob`

----

Yes, `Cause` should never return `nil` if it was passed something non `nil` so there is a bug there!  Do you have an example to trigger the problem?

```
--- a/fs/fserrors/error.go
+++ b/fs/fserrors/error.go
@@ -252,6 +252,10 @@ func ShouldRetry(err error) bool {
                return true
        }
 
+       // FIXME Handle this correctly, perhaps Cause should not ever return nil?
+       if err == nil {
+               return false
+       }
        // Check if it is a retriable error
        for _, retriableErr := range retriableErrors {
                if err == retriableErr {
```

----

And this one

```
+       // FIXME: Instead of having tight loop to check copy status, sleep and recheck?
+       // copyID := startCopy.CopyID()
```

Here is an example of how I did a similar thing in onedrive

https://github.com/ncw/rclone/blob/b9b9bce0db64abdfabc9fbdca26416e40400eb93/backend/onedrive/onedrive.go#L737-L738

I might be tempted to double the timeout each time up to a maximum but start much smaller.

How come the old code didn't have this?  Was it covered by the library or did I just miss it?

How long do copies normally take?

----

This is looking really good - looking forward to the next iteration.

sandeepkru commented:
Thanks for the suggestions and quick turnaround
> And for the _unsupported file (which doesn't exist yet, but you'll need it otherwise the builds will break).

Added as suggested and verified locally, everything compiles as expected
> can you prefix the line wiith vendor: rather than backend/azureblob

Done

> Yes, Cause should never return nil if it was passed something non nil so there is a bug there! Do you have an example to trigger the problem?
It is very easy to repro, copy operation or sync with new destination container always fails, below is the problem. Since target doesn't exist initially we throw "ContainerNotExist" storageError

In below code pkg/errors - I believe storageError returns causer interface but when this performs ``cause.Cause()`` it returns ``nil`` thus in ``fserror`` ShouldRetry causes segfault
https://github.com/ncw/rclone/blob/794e55de272aba98930fda2220d8c3d3b9f60673/vendor/github.com/pkg/errors/errors.go#L266

ShouldRetry here which would have called Cause gets err as nil and doing err.Error() causes segfault
https://github.com/ncw/rclone/blob/794e55de272aba98930fda2220d8c3d3b9f60673/fs/fserrors/error.go#L263

> Here is an example of how I did a similar thing in onedrive

Thanks, I added inline sleep instead of separating this as new function since this check is at copy operation level. Please let me know if you think I need to factor otherwise.

> How come the old code didn't have this? Was it covered by the library or did I just miss it?

Yes, client library was taking care of it earlier with tight loop
https://github.com/Azure/azure-sdk-for-go/blob/7971189ecf5a584b9211f2527737f94bb979644e/storage/copyblob.go#L160

> How long do copies normally take?

I did not time it explicitly with large blobs, for small blobs it was pretty quick. As per my understanding If it is with in azure I believe it should do at 80-100MB/s easily.

***
> I haven't attempted to review the code yet other than looking at the FIXMEs - I thought we should get it passing the tests first.

Agree, it was good idea and I hope tests pass now. I was trying to achieve code parity initially instead of doing any optimizations or refactoring as per latest SDK. Please suggest if you have any comments on code and I'll modify accordingly.

sandeepkru commented:
``` console
FAIL	github.com/ncw/rclone/backend/azureblob [build failed]
```
I see ``make quicktest`` fails in Travis, but when I tried locally it succeeded. Is it because of go version?

ncw commented:
The error in https://travis-ci.org/ncw/rclone/jobs/399383702 is complaining about goimports

```
find . -name \*.go | grep -v /vendor/ | xargs goimports -d | grep . ; test $? -eq 1
diff -u ./cmd/serve/webdav/webdav.go.orig ./cmd/serve/webdav/webdav.go
--- ./cmd/serve/webdav/webdav.go.orig	2018-07-03 00:59:34.732176391 +0000
+++ ./cmd/serve/webdav/webdav.go	2018-07-03 00:59:34.732176391 +0000
@@ -16,6 +16,7 @@
 	"github.com/ncw/rclone/vfs/vfsflags"
 	"github.com/spf13/cobra"
 	"golang.org/x/net/context" // switch to "context" when we stop supporting go1.8
+
 	"golang.org/x/net/webdav"
 )
 
diff -u ./fs/rc/rc.go.orig ./fs/rc/rc.go
--- ./fs/rc/rc.go.orig	2018-07-03 00:59:35.660176028 +0000
+++ ./fs/rc/rc.go	2018-07-03 00:59:35.660176028 +0000
@@ -12,6 +12,7 @@
 	"io"
 	"net/http"
 	_ "net/http/pprof" // install the pprof http handlers
+
 	"strings"
 
 	"github.com/ncw/rclone/cmd/serve/httplib"
make: *** [check] Error 1
```

`make check` locally will pick that up.  (You might need to do `make build_dep` first).  It is saying run your file through goimports.

sandeepkru commented:
Did a performance test - I did sync on a 10GiB file from azure vm to azure storage account with in same region and below are transfer details

Azure VM to Azure Storage 
```console
2018/07/03 23:36:07 DEBUG : file.txt: Uploading part 2560/2560 offset 9.996G/10G part size 4M
2018/07/03 23:36:08 INFO  : file.txt: Copied (new)
2018/07/03 23:36:08 INFO  : Waiting for deletions to finish
2018/07/03 23:36:08 INFO  :
Transferred:    10 GBytes (74.922 MBytes/s)
Errors:                 0
Checks:                 0
Transferred:            1
Elapsed time:     2m16.6s
```
Azure Storage to Azure VM
``` console
2018/07/03 23:39:33 INFO  : file.txt: Copied (new)
2018/07/03 23:39:33 INFO  : Waiting for deletions to finish
2018/07/03 23:39:33 INFO  :
Transferred:    10 GBytes (89.829 MBytes/s)
Errors:                 0
Checks:                 0
Transferred:            1
Elapsed time:     1m53.9s
```
***
One more test on Copy operation since you asked me earlier
From Azure Storage to Azure Storage with in same account 10GiB file
```console
Transferred:      0 Bytes (0 Bytes/s)
Errors:                 0
Checks:                 0
Transferred:            1
Elapsed time:       400ms
```
From azure vm file to azure storage - It is same as sync since underlying we perform multipart upload
```console
2018/07/03 23:45:58 DEBUG : file.txt: Uploading part 2560/2560 offset 9.996G/10G part size 4M
2018/07/03 23:45:58 INFO  : file.txt: Copied (new)
2018/07/03 23:45:58 INFO  :
Transferred:    10 GBytes (73.610 MBytes/s)
Errors:                 0
Checks:                 0
Transferred:            1
Elapsed time:     2m19.1s
```
Please let me know if you have any questions

sandeepkru commented:
> Sorry to do this to you, but I have merged #2118 which adds shared access signature support to the existing azure backend. It has been in progress for some time (since March) alas.

No worries, I was able to rebase and get to working form easily. Here is one problem though, I see new code has accepting credentials in three ways 1) Key 2) SAS URL 3) Connection String. Unfortunately, latest SDK doesn't support connection string, it has only two ways of accepting credentials
``NewAnonymusCredential`` or ``NewSharedAccessKeyCredential`` - Please refer to [GoDoc here](https://godoc.org/github.com/Azure/azure-storage-blob-go/2017-07-29/azblob#hdr-Credentials). 

To address this issue here are two options I've in my mind.

1.  Remove support for Connection String for now and open up an issue to new SDK team to support credentials through connection string. If they support, we can add this feature back to rclone.

2. Since new SDK seems to be designed to classify credentials only into two buckets, and in fact connection string is essentially a wrapper of storage account and key or some variation of sas_url, we can add our own parsing functionality on connection string and use appropriate credential mechanism based on content, so more or less implementing what exists in older SDK - https://github.com/Azure/azure-sdk-for-go/blob/fbe7db0e3f9793ba3e5704efbab84f51436c136e/storage/client.go#L221

Connection Strings can exist in various forms - [Official Connection String Documentation](https://docs.microsoft.com/en-us/azure/storage/common/storage-configure-connection-string), we should be able to support all these I believe.

First one is less work 😃 , but I'm more inclined doing second one,  though it involves more work. Please let me know what do you think and I can proceed accordingly.

sandeepkru commented:
**Update** - I followed up with new SDK team, apparently since there is no formal syntax defined on connection strings, they are planning to deprecate connection strings completely. Hence its been deprecated in latest go SDK, which also true in [latest java SDK](https://github.com/Azure/azure-storage-java-async/tree/c490f7a32915635eb2e2a379eff960addb5fa9cf/azure-storage/src/main/java/com/microsoft/azure/storage/blob). Considering this it seems wise to drop support on connection string.

ncw commented:
Great job on merging :-)

I just realised I fogot to merge the documentation I wrote for the 3 auth methods - I've done that here 71c974bf9af383947701e6bda9e27bc250f4d530. Can you pick that up in your PR too (with rebase) and edit appropriately!

I don't have a problem removing the Connection string method since this method hasn't been through a release yet or changing the authentication methods as long as the functionality works.

It seems like a good idea to remove the connection string method if the functionality is being deprecated anyway.

From the docs I posted (and the testing I did) it seems like a connection string is the only way to use an account wide SAS - the SAS_URL support seems only to work for a SAS tied to an individual bucket?  Do you think the new SDK works in a similar way? Or was that just an oddity of the old SDK?

ncw commented:
> I'm seeing issues with 72392a2 - MkDir takes container mutex and calls into list which needs same mutex, and I'm seeing deadlock because of this pattern. I'm not familiar if go supports reentrant locks or am I missing something here. Did you notice anything like this?

Go doesn't support re-entrant locks.  I think that code looks wrong - I don't think it can have ever worked :-(

> In my latest code, if create fails, I'm checking for container already exist error hence I think we might not need this code after all.

The point 72392a2 that patch is to avoid calling Mkdir to see if a container exists.  The reason for that being is that a SAS tied to a specific container can't call Mkdir.

The backend contract (which should be enforced by the integration tests) states that you should be able to call Mkdir on an existing directory (container) and it not do anything.  

I think doing  `rclone copy stuff azureblob:container` with a container based SAS URL will show the problem.

Rclone really only needs to do the list when using a container based SAS URL.

I'm happy for you to fix this up in any way you think fit!

> It looks like it was limitation in older SDK, I tried account level SAS URL and I'm able to perform all operations and I was able to run all tests against that remote successfully.

That is good news :-)

> I'm ready to send latest PR, I'll rebase and factor commits correctly and send an update once I resolve the first issue mentioned here.

Thank you that sounds great!

sandeepkru commented:
> Rclone really only needs to do the list when using a container based SAS URL.

I've changed not to allow any operations if sas url's container and provided containers do not match so this way commands like `rclone copy stuff azureblob:container` works correctly, if user replaces `container` with `invalidcontainer` an error is thrown at at initialization it self. 
I'm not sure if I understood your concern correctly or not. Please let me know and I can fix if you think it is not a good idea.

ncw commented:
Good progress :-) It looks like we are nearly there!

I've been ran the intergration tests locally and found a few issues. (`cd backend/azureblob ; go test -v` - You'll need to make a `TestAzureBlob` remote.)

The first test caused a panic

```
=== RUN   TestIntegration/TestObjectOpenRange
panic: The range count must be >= 0 [recovered]
	panic: The range count must be >= 0
```

This can be fixed like this

```patch
diff --git a/backend/azureblob/azureblob.go b/backend/azureblob/azureblob.go
index 2e5efe24..049e353f 100644
--- a/backend/azureblob/azureblob.go
+++ b/backend/azureblob/azureblob.go
@@ -901,16 +901,10 @@ func (o *Object) Open(options ...fs.OpenOption) (in io.ReadCloser, err error) {
 	for _, option := range options {
 		switch x := option.(type) {
 		case *fs.RangeOption:
-			start, end := x.Start, x.End
-			if end < 0 {
-				end = 0
+			offset, count = x.Decode(o.size)
+			if count < 0 {
+				count = o.size - offset
 			}
-			if start < 0 {
-				start = o.size - end
-				end = 0
-			}
-			offset = start
-			count = end - start
 		case *fs.SeekOption:
 			offset = x.Offset
 		default:
```

Then we get

```
--- FAIL: TestIntegration (2.81s)
    --- FAIL: TestIntegration/TestFsPutError (0.07s)
    	fstests.go:394: 
    			Error Trace:	fstests.go:394
    			Error:      	Expected value not to be nil.
    			Test:       	TestIntegration/TestFsPutError
    	fstests.go:397: 
    			Error Trace:	fstests.go:397
    			Error:      	Expected nil, but got: &azureblob.Object{fs:(*azureblob.Fs)(0xc420434750), remote:"hello? sausage/êé/Hello, 世界/ \" ' @ < > & ? + ≠/z.txt", modTime:time.Time{wall:0x756b5b3, ext:63116769910, loc:(*time.Location)(nil)}, md5:"/BP1D+k3XvseD7AM7LWk4A==", size:50, mimeType:"text/plain; charset=utf-8", accessTier:"", meta:map[string]string{"mtime":"2001-02-03T04:05:10.123123123Z"}}
    			Test:       	TestIntegration/TestFsPutError
    	fstests.go:398: 
    			Error Trace:	fstests.go:398
    			Error:      	Not equal: 
    			            	expected: *errors.fundamental(object not found)
    			            	actual  : <nil>(<nil>)
    			Test:       	TestIntegration/TestFsPutError
    --- FAIL: TestIntegration/TestObjectSetModTime (0.03s)
    	fstest.go:121: 
    			Error Trace:	fstest.go:121
    			            				fstests.go:850
    			            				fstests.go:884
    			Error:      	Should be true
    			Test:       	TestIntegration/TestObjectSetModTime
    			Messages:   	file name.txt: Modification time difference too big |-95170h10m10.5s| > 1ns (2001-02-03 04:05:06.499999999 +0000 UTC vs 2011-12-13 14:15:16.999999999 +0000 UTC) (precision 1ns)
    --- FAIL: TestIntegration/TestObjectPurge (0.04s)
    	fstests.go:1131: 
    			Error Trace:	fstests.go:1131
    			Error:      	An error is expected but got nil.
    			Test:       	TestIntegration/TestObjectPurge
    			Messages:   	Expecting error after on second purge
FAIL
```

TestFsPutError checks that errors on the input to the blob reader make it back out.  So this creates an input which will give an error and checks an error comes back. So there is an error check missing somewhere but it might be in the SDK...

The TestSetModTime is caused by the SetModTime call not writing back the metadata to the blob - there is just a bit of code missing in the method by the look of it!

TestObjectPurge is expecting an error if you try to delete an already deleted container.

Can you have a go at fixing these up?

Thanks

Nick

sandeepkru commented:
> I've been ran the integration tests locally and found a few issues

My bad, I was running integration tests from fs/sync instead of backend/azureblob. I see similar issues as well.

- ``TestObjectOpenRange`` - Thanks, your suggested change fixed this.

- ``TestFsPutError`` is not actually failing to put object though it has all random characters in the path, hence later when it tries to get object, the operation is getting successful. As per my understanding intention here is to make first operation to fail as well? If that's the case, I may have to change path to make initial put operation to fail.
 
- ``TestObjectSetModTime`` - I wasn't setting on remote since update operation does that, but looks like in this test expectation is to set on remote. I tried and having some other issues and currently debugging it.

- ``TestObjectPurge`` - Working on it.

> Can you have a go at fixing these up?

Planning to send it soon.

ncw commented:
> `TestFsPutError` is not actually failing to put object though it has all random characters in the path, hence later when it tries to get object, the operation is getting successful. As per my understanding intention here is to make first operation to fail as well? If that's the case, I may have to change path to make initial put operation to fail.

What `TestFsPutError` is gives a reader the backend method `Put` which returns an error.  This should cause the `Put` method to return an error which it did.  However rclone expects the object not to be created in that case which is the part of the test that is failing I think.

sandeepkru commented:
> What TestFsPutError is gives a reader the backend method Put which returns an error. This should cause the Put method to return an error which it did. However rclone expects the object not to be created in that case which is the part of the test that is failing I think.

I investigated more on this, here is what I observed, I understood the test expectation, you are correct in this case it is not failing to create object instead it is succeeded thus later assertions in test are failing.

I looked into client SDK as shown here [azblob/highlevel.go#L493](https://github.com/Azure/azure-storage-blob-go/blob/f0a732ea9441c73511367d5861da6bda664ffabd/2017-07-29/azblob/highlevel.go#L493) though the `err` is not `nil` they are still populating buffer and uploading contents to azure storage and at the end they are not really returning error to user [highlevel.go#L509](https://github.com/Azure/azure-storage-blob-go/blob/f0a732ea9441c73511367d5861da6bda664ffabd/2017-07-29/azblob/highlevel.go#L509)

Where as in older SDK it was returning error to user, it looks like a bug in client version of SDK. I can open up an issue with that team. Please let me know how do you want to proceed meanwhile.

- `TestObjectSetModTime` - Fixed this now to behave as expected.
- `TestObjectPurge` - This was failing due to how `delete` functionality implemented at `azure` end, successive delete calls will not fail up to 30 - 45 seconds though first `delete` call deletes specified `object`. In old code on `deleteContainer`,  first `Exists` call was performed hence it was working correctly. In latest SDK, they removed `Exists` on container level hence latest code changes observed this failure. However, I changed to do `GerProperties` in place of `Exists` and it is working as expected (I am not very happy about this pattern but looks like this is best option available if we would like to know object is really deleted or not).

***
On different note, I see they published latest `2018-03-28` SDK version recently and I believe it is stable, shall I go ahead and pull this version along with my changes?

> Good progress :-) It looks like we are nearly there!

Thanks, except that one blocking test, it seems almost ready :). I'm ready to send new iteration based on your response.

ncw commented:
> I looked into client SDK as shown here azblob/highlevel.go#L493 though the err is not nil they are still populating buffer and uploading contents to azure storage and at the end they are not really returning error to user highlevel.go#L509

That looks like a bug to me - I would report it to the SDK team I think.  Maybe even send a PR if you can think of a neat way of fixing it!

As a workaround what you can do is delete the object if you received an error (some of the other backends do this to stop incomplete objects lying around, eg https://github.com/ncw/rclone/blob/master/backend/local/local.go#L789 )

> TestObjectPurge - This was failing due to how delete functionality implemented at azure end, successive delete calls will not fail up to 30 - 45 seconds though first delete call deletes specified object. In old code on deleteContainer, first Exists call was performed hence it was working correctly. In latest SDK, they removed Exists on container level hence latest code changes observed this failure. However, I changed to do GerProperties in place of Exists and it is working as expected (I am not very happy about this pattern but looks like this is best option available if we would like to know object is really deleted or not).

Ah, interesting.  It sounds like a reasonable workaround. 

> On different note, I see they published latest 2018-03-28 SDK version recently and I believe it is stable, shall I go ahead and pull this version along with my changes?

That sounds like a good idea - will it be a disruptive change?

sandeepkru commented:
> As a workaround what you can do is delete the object if you received an error (some of the other backends do this to stop incomplete objects lying around

Thanks for the pointer, but in this case client library is masking error completely, we wouldn't even know if was uploaded partially (in this particular case, it is uploading completely) since error is not returned to the user.

> That sounds like a good idea - will it be a disruptive change?

This will not be disruptive, it is as simple as updating api version in previous SDK.

> That looks like a bug to me - I would report it to the SDK team I think. Maybe even send a PR if you can think of a neat way of fixing it!

I can do this, for now is there a way to disable this test on this backend alone? I'm not sure how long it takes to get fix in mainstream.

ncw commented:
> Thanks for the pointer, but in this case client library is masking error completely, we wouldn't even know if was uploaded partially (in this particular case, it is uploading completely) since error is not returned to the user.

Yes you are right, I mis-read the line numbers in the test failure.

> I can do this, for now is there a way to disable this test on this backend alone? I'm not sure how long it takes to get fix in mainstream.

I'd rather get what you have done merged into rclone now, then I'm happy for this one test to fail while we work on an upstream fix which probably won't take too long.

sandeepkru commented:
> I'm happy for this one test to fail while we work on an upstream fix which probably won't take too long

I opened an issue https://github.com/Azure/azure-storage-blob-go/issues/52, hopefully they will have fix soon, otherwise I'll think about correct fix and send out PR to them.

>> That sounds like a good idea - will it be a disruptive change?

>This will not be disruptive, it is as simple as updating api version in previous SDK.

I was hoping this would be trivial, unfortunately they changed few items,  I was able to fix those and pull latest version and ran all tests seems everything is fine except that one test case.

It is almost ready except this one `FIXME` on [Accounting](https://github.com/sandeepkru/rclone/blob/5032d8b7f0cb80d59f42729abbcdb65929bf0d45/backend/azureblob/azureblob.go#L1036). Current code bypasses wrap and unwrap functionalities on accounting. I see that wrap returns `io.Reader` but latest SDK [StageBlock](https://github.com/sandeepkru/rclone/blob/5032d8b7f0cb80d59f42729abbcdb65929bf0d45/backend/azureblob/azureblob.go#L1077) expects `io.ReadSeeker`, I do not know an easier way to achieve this without any additional copies of buffers. Please let me know if you have any ideas on this and I'll change based on your suggestions.

There is also some duplication with `decodeMetada`, which I was not very happy about but with latest SDK, I couldn't avoid that. At this point of time, it is closer to the older code in working condition, we can iterate on this optimizing further once it makes into mainstream.

ncw commented:
> It is almost ready except this one FIXME on Accounting. Current code bypasses wrap and unwrap functionalities on accounting. I see that wrap returns io.Reader but latest SDK StageBlock expects io.ReadSeeker, I do not know an easier way to achieve this without any additional copies of buffers. Please let me know if you have any ideas on this and I'll change based on your suggestions.

The point of the accounting changes are so that the accounting knows when the data is read from the buffer and transmitted.  SInce the blocks on Azure blob are relatively small (4MB I think) then this will not be too noticeable if we can't fix it.  However you could fix it something like this:

```diff
diff --git a/backend/azureblob/azureblob.go b/backend/azureblob/azureblob.go
index 2e5efe24..4e50c0eb 100644
--- a/backend/azureblob/azureblob.go
+++ b/backend/azureblob/azureblob.go
@@ -23,6 +23,7 @@ import (
 
 	"github.com/Azure/azure-storage-blob-go/2017-07-29/azblob"
 	"github.com/ncw/rclone/fs"
+	"github.com/ncw/rclone/fs/accounting"
 	"github.com/ncw/rclone/fs/config"
 	"github.com/ncw/rclone/fs/config/flags"
 	"github.com/ncw/rclone/fs/fserrors"
@@ -955,6 +956,12 @@ func init() {
 	}
 }
 
+// readSeeker joins an io.Reader and an io.Seeker
+type readSeeker struct {
+	io.Reader
+	io.Seeker
+}
+
 // uploadMultipart uploads a file using multipart upload
 //
 // Write a larger blob, using CreateBlockBlob, PutBlock, and PutBlockList.
@@ -1011,10 +1018,9 @@ func (o *Object) uploadMultipart(in io.Reader, size int64, blob *azblob.BlobURL,
 	ctx := context.Background()
 	ac := azblob.LeaseAccessConditions{} // Use default lease access conditions
 
-	// FIXME - Accounting
 	// unwrap the accounting from the input, we use wrap to put it
 	// back on after the buffering
-	// in, wrap := accounting.UnWrap(in)
+	in, wrap := accounting.UnWrap(in)
 
 	// Upload the chunks
 	remaining := size
@@ -1055,7 +1061,10 @@ outer:
 			fs.Debugf(o, "Uploading part %d/%d offset %v/%v part size %v", part+1, totalParts, fs.SizeSuffix(position), fs.SizeSuffix(size), fs.SizeSuffix(chunkSize))
 
 			err = o.fs.pacer.Call(func() (bool, error) {
-				_, err = blockBlobURL.StageBlock(ctx, blockID, bytes.NewReader(buf), ac)
+				bufferReader := bytes.NewReader(buf)
+				wrappedReader := wrap(bufferReader)
+				rs := readSeeker{wrappedReader, bufferReader}
+				_, err = blockBlobURL.StageBlock(ctx, blockID, rs, ac)
 				return o.fs.shouldRetry(err)
 			})
 ```

> There is also some duplication with decodeMetada, which I was not very happy about but with latest SDK, I couldn't avoid that. At this point of time, it is closer to the older code in working condition, we can iterate on this optimizing further once it makes into mainstream.

Good plan.

sandeepkru commented:
>  However you could fix it something like this:

Thanks for sharing this, I've added now and everything works fine.

>  I haven't attempted to review the code yet other than looking at the FIXMEs - I thought we should get it passing the tests first.

I believe we are finally at place to get this reviewed 😄 , please review changes and provide feedback, I'll modify based on your suggestions and hopefully we can merge these changes soon.

sandeepkru commented:
Update:

> I'd rather get what you have done merged into rclone now, then I'm happy for this one test to fail while we work on an upstream fix which probably won't take too long.

Azure team shared a private with me  fixing this Azure/azure-storage-blob-go#52, and I ran integration tests with it, now all tests pass including `TestFsPutError`. Hope, this fix will come soon to release.

sandeepkru commented on backend/azureblob/azureblob.go:
```diff
@@ -190,53 +189,54 @@ func NewFs(name, root string) (fs.Fs, error) {
-	connectionString := config.FileGet(name, "connection_string")
 	sasURL := config.FileGet(name, "sas_url")
-	endpoint := config.FileGet(name, "endpoint", storage.DefaultBaseURL)
+	endpoint := config.FileGet(name, "endpoint", "blob.core.windows.net")
```
Done

sandeepkru commented on backend/azureblob/azureblob.go:
```diff
@@ -342,32 +337,39 @@ func (f *Fs) list(dir string, recurse bool, maxResults uint, fn listFn) error {
+
 		if err != nil {
-			if storageErr, ok := err.(storage.AzureStorageServiceError); ok && storageErr.StatusCode == http.StatusNotFound {
+			// Check http error code along with service code, current SDK doesn't populate service code correctly sometimes
```
I have not yet, but I do have a easier repro for this, will report it today.

sandeepkru commented on backend/azureblob/azureblob.go:
```diff
@@ -784,11 +785,41 @@ func (o *Object) decodeMetaData(info *storage.Blob) (err error) {
+}
+
+func (o *Object) decodeMetaDataFromBlob(info *azblob.BlobItem) (err error) {
+	o.md5 = string(info.Properties.ContentMD5[:])
```
It is base64 encoded, today there is this inconsistency with client SDK, with `BlobProperties` `ContentMD5` is returned as `base64 encoded []byte`, where as in `azBlob.BlobGetPropertiesResponse` exposes a `getter` on `ContentMD5` which returns a `base64 decoded string` as used in `decodeMetaDataFromPropertiesResponse`. I reported this inconsistency to SDK team, hope they will fix it in near future.

sandeepkru commented on backend/azureblob/azureblob.go:
```diff
@@ -784,11 +785,41 @@ func (o *Object) decodeMetaData(info *storage.Blob) (err error) {
+//  o.md5
+//  o.meta
+func (o *Object) decodeMetaDataFromPropertiesResponse(info *azblob.BlobGetPropertiesResponse) (err error) {
+	// FIXME - Client library returns MD5 as base64 decoded string, object struct should be changed
```
On second thought, I think it is good idea to maintain it as string, either way at some place we have to convert this `[]byte` or `string` due to the inconsistencies in SDK as explained in above comment. Also, maintaining it as string would be much easier to debug hence I updated this comment.

sandeepkru commented on fs/fserrors/error.go:
```diff
@@ -252,6 +252,10 @@ func ShouldRetry(err error) bool {
 		return true
 	}
 
+	// FIXME Handle this correctly, perhaps Cause should not ever return nil?
```
Done, thanks for this fix.

sandeepkru commented:
> That is looking really good! I put a few very small things inline.

Thank you!

ncw commented:
I've merged that now - Thank you very much for your hard work - much appreciated :-)

ncw commented:
BTW I found a small bug in this when I was rebasing the config branch:

b0d1fa1d6bba5ea6b78b8af011d20791574cb2ed

It would give nil pointer errors if the error returned wasn't of the type asserted.

sandeepkru commented:
> I've merged that now - Thank you very much for your hard work - much appreciated :-)

Thanks for your guidance and all the pointers, I'll get started on other #2091.

> It would give nil pointer errors if the error returned wasn't of the type asserted.

My bad, thanks for fixing it.

ncw commented:
@sandeepkru great!  Note I'm going to merge some changes to all the backends shortly to unify the configuration methods...

sandeepkru commented:
> Note I'm going to merge some changes to all the backends shortly to unify the configuration methods...

Thanks for the heads up Nick, I think, its best that I'll wait until your changes are ready. Meanwhile, I'll work on bugs related `GetModifyWindow` and `ServiceCode` not getting populated correctly.

Do you have any `work in progress` reference to understand how configuration methods are getting changed?

sandeepkru commented:
I was away and couldn't get a chance to look at this earlier, I opened an issue [#58](https://github.com/Azure/azure-storage-blob-go/issues/58) with Azure SDK team on error handling. I  pulled your latest changes, will send out soon a small write up on my approach to enable `SetBlobTier` feature on azure backend

ncw commented:
@sandeepkru great - thanks :-)
